lib/slurm_squeue_client.rb

```diff
@@ -61,7 +61,73 @@ def sinfo
     
     s.success? ? @sinfo = o : raise(CommandFailed, e)
   end
-  
+
+  # Return length of GRES field from SLURM
+  # @return [Integer] gres length
+  def gres_length
+    return @gres_length if defined?(@gres_length)
+
+    Open3.pipeline_rw "sinfo -o '%G' | awk '{ print length }'", "sort -n", "tail -1" do |stdin,stdout|
+      stdin.write stdout
+      stdin.close
+      @gres_length = stdout.read.to_i
+    end
+  end
+
+  # Parse and return total number of GPU nodes in a SLURM cluster
+  # @return [Integer] number of GPU nodes
+  def gpu_nodes
+    return @available_gpu_nodes if defined?(@available_gpu_nodes)
+
+    Open3.pipeline_rw "sinfo -N -h -a --Format='nodehost,gres:#{@gres_length}'", 'uniq', 'grep gpu:', 'wc -l' do |stdin, stdout|
+      stdin.write stdout
+      stdin.close
```

Comment:
In this case it seems like you would want to use `Open3.pipeline_r` instead of `Open3.pipeline_rw` as you don't need to write anything to stdin of the sinfo command, you just need to read stdout.